server/lib/parties/party-service.ts

```diff
@@ -0,0 +1,228 @@
+
+  invite(leader: PartyUser, leaderClientId: string, invites: PartyUser[]): Readonly<PartyRecord> {
+    const leaderClientSockets = this.getClientSockets(leader.id, leaderClientId)
+
+    let party = this.getClientParty(leaderClientSockets)
+    if (party) {
+      if (party.leaderId !== leader.id) {
+        throw new PartyServiceError(
+          PartyServiceErrorCode.InsufficientPermissions,
+          'Only party leader can invite people',
+        )
+      }
+
+      for (const invite of invites) {
+        party.invites.set(invite.id, invite)
+      }
+
+      this.publishToParty(party.id, {
+        type: 'invite',
+        invites,
+      })
+    } else {
+      const partyId = cuid()
+      party = {
+        id: partyId,
+        invites: new Map(invites.map(i => [i.id, i])),
+        members: new Map([[leader.id, leader]]),
+        leaderId: leader.id,
+      }
+
+      this.parties.set(partyId, party)
+      this.clientSocketsToPartyId.set(leaderClientSockets, partyId)
+      this.subscribeToParty(leaderClientSockets, party)
+    }
+
+    invites
+      .map(i => this.getUserSockets(i.name))
+      .forEach(userSockets => {
+        userSockets.subscribe(
+          getInvitesPath(party!.id),
+          () => ({
+            type: 'invite',
+            from: party!.leaderId,
+          }),
+          () => {
+            // TODO(2Pac): Handle user quitting; need to keep a map of user -> invites?
+          },
+        )
+      })
+
+    return party
+  }
+
+  removeInvite(partyId: string, target: PartyUser, leader?: PartyUser) {
+    const party = this.parties.get(partyId)
+    if (!party) {
+      throw new PartyServiceError(PartyServiceErrorCode.PartyNotFound, 'Party not found')
+    }
+
+    if (leader && leader.id !== party.leaderId) {
+      throw new PartyServiceError(
+        PartyServiceErrorCode.InsufficientPermissions,
+        'Only party leaders can remove invites to other people',
+      )
+    }
+
+    // TODO(2Pac): Check if the target is invited first before trying to do any of the below?
+
+    party.invites.delete(target.id)
+    this.publishToParty(partyId, {
+      type: 'decline',
+      target,
+    })
+
+    // TODO(2Pac): Do we want to remove the party record if there's only one member left in a party
+    // and no outstanding invites?
+
+    const targetUserSockets = this.getUserSockets(target.name)
+    targetUserSockets.unsubscribe(getInvitesPath(partyId))
+  }
+
+  acceptInvite(partyId: string, user: PartyUser, clientId: string) {
+    const clientSockets = this.getClientSockets(user.id, clientId)
+
+    const party = this.parties.get(partyId)
+    if (!party) {
+      throw new PartyServiceError(PartyServiceErrorCode.PartyNotFound, 'Party not found')
+    }
+
+    if (party.members.size >= MAX_PARTY_SIZE) {
+      throw new PartyServiceError(PartyServiceErrorCode.PartyFull, 'Party is full')
+    }
+
+    // TODO(2Pac): Check if the target is invited first before trying to do any of the below?
```

Comment:
yes, obv :) don't want to let people join themselves into random other parties they weren't invited to